nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DetectDuplicateRecord.java

```diff
@@ -0,0 +1,620 @@
+
+                if(filter.contains(recordHash)) {
+                    duplicatesWriter.write(record);
+                } else {
+                    nonDuplicatesWriter.write(record);
+                }
+
+                filter.put(recordHash);
+            }
+
+            final boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).isSet()
+                    ? context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean()
+                    : true;
+
+
+            // Route Non-Duplicates FlowFile
+            final WriteResult nonDuplicatesWriteResult = nonDuplicatesWriter.finishRecordSet();
+            Map<String, String> attributes = new HashMap<>();
+            attributes.putAll(nonDuplicatesWriteResult.getAttributes());
+            attributes.put("record.count", String.valueOf(nonDuplicatesWriteResult.getRecordCount()));
+            attributes.put(CoreAttributes.MIME_TYPE.key(), nonDuplicatesWriter.getMimeType());
+            nonDuplicatesFlowFile = session.putAllAttributes(nonDuplicatesFlowFile, attributes);
+            logger.info("Successfully found {} unique records for {}", new Object[] {nonDuplicatesWriteResult.getRecordCount(), nonDuplicatesFlowFile});
+
+            if(!includeZeroRecordFlowFiles && nonDuplicatesWriteResult.getRecordCount() == 0) {
+                session.remove(nonDuplicatesFlowFile);
+            } else {
+                session.transfer(nonDuplicatesFlowFile, REL_NON_DUPLICATE);
+            }
+
+            // Route Duplicates FlowFile
+            final WriteResult duplicatesWriteResult = duplicatesWriter.finishRecordSet();
+            attributes.clear();
+            attributes.putAll(duplicatesWriteResult.getAttributes());
+            attributes.put("record.count", String.valueOf(duplicatesWriteResult.getRecordCount()));
+            attributes.put(CoreAttributes.MIME_TYPE.key(), duplicatesWriter.getMimeType());
+            nonDuplicatesFlowFile = session.putAllAttributes(nonDuplicatesFlowFile, attributes);
+            logger.info("Successfully found {} duplicate records for {}", new Object[] {duplicatesWriteResult.getRecordCount(), nonDuplicatesFlowFile});
+
+            if(!includeZeroRecordFlowFiles && duplicatesWriteResult.getRecordCount() == 0) {
+                session.remove(duplicatesFlowFile);
+            } else {
+                session.transfer(duplicatesFlowFile, REL_DUPLICATE);
+            }
+
+            session.adjustCounter("Records Processed",
+                    nonDuplicatesWriteResult.getRecordCount() + duplicatesWriteResult.getRecordCount(), false);
+
+            if(shouldCacheIdentifier) {
+                CacheValue cacheValue = new CacheValue(serializableFilter, now);
+                cache.put(cacheKey, cacheValue, keySerializer, cacheValueSerializer);
+            }
+
+        } catch (final SchemaNotFoundException e) {
+            throw new ProcessException(e.getLocalizedMessage(), e);
```

Comment:
I think we should route to failure on both of these because you can cause a flow to hang this way.

Reply:
I used `AbstractRecordProcessor` as a guide for writing this processor and it [contained this same code here](https://github.com/apache/nifi/blob/36c0a99e91c492329c0df69ee4ae961e57295f84/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AbstractRecordProcessor.java#L167). It's a shame I couldn't inherit from that class which I initially tried doing but this processor has to do additional things at the file level which is why I moved away from it. Regardless, does that mean `AbstractRecordProcessor` should also be updated?

Reply:
I think I see what you mean. Perhaps moving the session transfer calls outside the try block would suffice?

```
session.transfer(nonDuplicatesFlowFile, REL_NON_DUPLICATE);
session.transfer(duplicatesFlowFile, REL_DUPLICATE);
```